Source/Core/GoogleEarthEnterpriseMetadata.js

```diff
@@ -0,0 +1,619 @@
+/*global define*/
+define([
+    './appendForwardSlash',
+    './defaultValue',
+    './defined',
+    './defineProperties',
+    './DeveloperError',
+    './loadArrayBuffer',
+    './RuntimeError',
+    '../ThirdParty/pako_inflate',
+    '../ThirdParty/when'
+], function(
+    appendForwardSlash,
+    defaultValue,
+    defined,
+    defineProperties,
+    DeveloperError,
+    loadArrayBuffer,
+    RuntimeError,
+    pako,
+    when) {
+    'use strict';
+
+    // Bitmask for checking tile properties
+    var childrenBitmasks = [0x01, 0x02, 0x04, 0x08];
+    var anyChildBitmask = 0x0F;
+    var cacheFlagBitmask = 0x10; // True if there is a child subtree
+    //var vectorDataBitmask = 0x20;
+    var imageBitmask = 0x40;
+    var terrainBitmask = 0x80;
+
+    // Datatype sizes
+    var sizeOfUint16 = Uint16Array.BYTES_PER_ELEMENT;
+    var sizeOfInt32 = Int32Array.BYTES_PER_ELEMENT;
+    var sizeOfUint32 = Uint32Array.BYTES_PER_ELEMENT;
+
+    function isBitSet(bits, mask) {
+        return ((bits & mask) !== 0);
+    }
+
+    function TileInformation(bits, cnodeVersion, imageryVersion, terrainVersion) {
+        this._bits = bits;
+        this.cnodeVersion = cnodeVersion;
+        this.imageryVersion = imageryVersion;
+        this.terrainVersion = terrainVersion;
+        this.ancestorHasTerrain = false; // Set it later once we find its parent
+        this.terrainState = 0; // UNKNOWN
+    }
+
+    TileInformation.prototype.setParent = function(parent) {
+        this.ancestorHasTerrain = parent.ancestorHasTerrain || this.hasTerrain();
+    };
+
+    TileInformation.prototype.hasSubtree = function() {
+        return isBitSet(this._bits, cacheFlagBitmask);
+    };
+
+    TileInformation.prototype.hasImagery = function() {
+        return isBitSet(this._bits, imageBitmask);
+    };
+
+    TileInformation.prototype.hasTerrain = function() {
+        return isBitSet(this._bits, terrainBitmask);
+    };
+
+    TileInformation.prototype.hasChildren = function() {
+        return isBitSet(this._bits, anyChildBitmask);
+    };
+
+    TileInformation.prototype.hasChild = function(index) {
+        return isBitSet(this._bits, childrenBitmasks[index]);
+    };
+
+    TileInformation.prototype.getChildBitmask = function(index) {
+        return this._bits && anyChildBitmask;
+    };
+
+    GoogleEarthEnterpriseMetadata.TileInformation = TileInformation;
+
+    /**
+     * Provides metadata using the Google Earth Enterprise REST API. This is used by the
+     *
+     * @alias GoogleEarthEnterpriseMetadata
```

Comment:
I had this setup where we had a static list of servers, so providers would automatically share these objects. I then had to add a destroy method to the providers so we could delete metadata objects that weren't used, that were called when they were removed. This broke a whole bunch of Sandcastle examples as once a provider was removed it could never be reused.